test/end2end_test.go

```diff
@@ -7032,6 +7032,84 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 	}()
 }
 
+// TestClientTransportRestartsAfterStreamIdExhausted tests that the client transport
+// drains and restarts when next stream ID exceeds MaxStreamID. This test also
+// verifies that subsequent RPCs use a new client transport and the old transport
+// is closed.
+func (s) TestClientTransportRestartsAfterStreamIdExhausted(t *testing.T) {
+	// overriding MaxStreamIDForTesting.
+	originalMaxStreamID := transport.MaxStreamIDForTesting
+	transport.MaxStreamIDForTesting = 5
+	defer func() {
+		transport.MaxStreamIDForTesting = originalMaxStreamID
+	}()
+
+	// setting up StubServer.
+	s := grpc.NewServer()
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 10; i++ {
+				m := &grpc.PreparedMsg{}
+				stream.SendMsg(m)
+			}
+			return nil
+		},
+	}
+	testpb.RegisterTestServiceServer(s, ss)
```

Comment:
It's unfortunate you have to have a ton of the boilerplate that `StubServer` is intended to help with.

Two options:

1. use `WithContextDialer` on the client to detect when we're (re)connecting.
2. Add some stuff to `StubServer` to facilitate wrapping the listener.

This will save you manually creating the server, registering the service, dialing the client, wrapping the client with the grpc stub, and dealing with extra shutdown tasks.  The resulting test code will be the more interesting code which will make it easier to read and understand what it's doing that's unique.